### DISCUSSION TITLE
Embedding the NATS server the right way

### DISCUSSION THREAD

**Discussion by maxpert** [nats-io/nats-server#4279] [Q&A]
Hey folks I've been struggling to embed NATS into my open source project. Here is [code on](https://github.com/maxpert/marmot/blob/master/stream/embedded_nats.go#L40-L76) how I am embedding NATs into Marmot. Now when I supply configuration file:

```
accounts: {
    STREAMERS: {
        jetstream: enabled
        users: [
            {user: marmot, password: PassW0rd12345}
        ]
    },
    SYS: {
        users: [
            {user: admin, password: PassW0rd123}
        ]
    },
}
system_account: SYS

cluster {
  name: we4u
  listen: localhost:4248
  routes = []
}

jetstream: enabled
jetstream {
    store_dir: "/tmp/nats1/stream"
    max_mem: 32M
    max_file: 1G
}
```

Launching server I get error:
```
2023/06/29 07:25:07.624997 [INF] Starting nats-server
2023/06/29 07:25:07.625020 [INF]   Version:  2.9.19
2023/06/29 07:25:07.625025 [INF]   Git:      [not set]
2023/06/29 07:25:07.625033 [INF]   Cluster:  we4u
2023/06/29 07:25:07.625037 [INF]   Name:     marmot-node-1
2023/06/29 07:25:07.625041 [INF]   Node:     OWL7P9aV
2023/06/29 07:25:07.625045 [INF]   ID:       NAQSEHD2K7G6ZQ6ARBMCT75LAQOTA7PVI5CLL6L5FIR65YNX5AX5BMVZ
2023/06/29 07:25:07.625065 [WRN] Plaintext passwords detected, use nkeys or bcrypt
2023/06/29 07:25:07.625088 [INF] Using configuration file: ./dist/n1-nats.conf
2023/06/29 07:25:07.625616 [INF] Starting JetStream
2023/06/29 07:25:07.626034 [INF]     _ ___ _____ ___ _____ ___ ___   _   __  __
2023/06/29 07:25:07.626044 [INF]  _ | | __|_   _/ __|_   _| _ \ __| /_\ |  \/  |
2023/06/29 07:25:07.626047 [INF] | || | _|  | | \__ \ | | |   / _| / _ \| |\/| |
2023/06/29 07:25:07.626050 [INF]  \__/|___| |_| |___/ |_| |_|_\___/_/ \_\_|  |_|
2023/06/29 07:25:07.626053 [INF]
2023/06/29 07:25:07.626055 [INF]          https://docs.nats.io/jetstream
2023/06/29 07:25:07.626058 [INF]
2023/06/29 07:25:07.626060 [INF] ---------------- JETSTREAM ----------------
2023/06/29 07:25:07.626077 [INF]   Max Memory:      30.52 MB
2023/06/29 07:25:07.626082 [INF]   Max Storage:     953.67 MB
2023/06/29 07:25:07.626085 [INF]   Store Directory: "/tmp/nats1/stream/jetstream"
2023/06/29 07:25:07.626089 [INF] -------------------------------------------
2023/06/29 07:25:07.626796 [INF] Starting JetStream cluster
2023/06/29 07:25:07.626811 [FTL] Can't start JetStream: JetStream cluster requires configured routes or solicited leafnode for the system account
```

Here is what I want Marmot to be able to do:
 - Launch a node with a NATS server as primary node
 - Add nodes to cluster that can have cluster peers
 - For quick start no NATs configuration and sane defaults so that I can pass only cluster address for primary and peer nodes as part of CLI. People can later use configuration file later to do more complex configurations. 

I am looking for expertise here to help me make NATs part of Marmot and really easy to get up and running while allowing complex configurations via nats config file if required.

**Comment by marcchane412** [comment]


**Reply by ripienaar** [reply]
Thanks. Yes it was spam. I removed it.

**Comment by mtfash** [comment]
@maxpert Did you find the solution?

**Comment by tehsphinx** [comment]
For anyone finding this and looking on how to solve this here the 2 options that worked for me:
- If it is a single node NATS cluster, removing the `cluster` section from the configuration solves this. That way there is no need for routes either.
- If having a cluster, the routes (addresses) of the other NATS servers need to be added to the `cluster` section as can be seen in this example on [JetStream Clusting](https://docs.nats.io/running-a-nats-service/configuration/clustering/jetstream_clustering).

I didn't dig deeper into the mention of leaf nodes in the error message.

**Comment by mrtdeh** [comment]
When using JetStream in NATS, the `cluster routes` field should never be empty. However, there may be situations where you only have a single node in the cluster. In such cases, you can specify the current node's address in the routes field. Don't worry - this won't cause any issues. In the configuration, it would look like this:

```bash
...
routes = ["nats://localhost:4248"]
...
```
### Important note: When operating in cluster mode or using gateways, the routes field must contain actual values and cannot be empty for JetStream to function.
